### PR TITLE
loosen constraints on solver signatures

### PR DESCRIPTION
**Pull request by froystig**
When defining implicit derivatives, allow solver functions to have whatever signature, including keyword arguments (whether explicit or via a `**kwargs` binder). If differentiated, enforce that the call is consistent with the signature of the optimality function or the fixed-point function, for `custom_root` and `custom_fixed_point` derivatives respectively.

### CODE REVIEW

File: jaxopt/_src/implicit_diff.py [hunk 1]
```diff
@@ -120,29 +120,124 @@ def root_jvp(optimality_fun: Callable,
...
+
+      ba_args, ba_kwargs, map_back = _signature_bind_and_match(
+          optimality_signature, *args, **kwargs)
+      if ba_kwargs:
+        raise TypeError(
+            "keyword arguments to solver_fun could not be resolved to "
+            f"positional arguments based on the signature {signature}. This "
+            "can happen under custom_root if optimality_fun takes catch-all "
+            "**kwargs, or under custom_fixed_point if fixed_point_fun takes "
+            "catch-all **kwargs, both of which are currently unsupported.")
```
Comment (froystig) [inline jaxopt/_src/implicit_diff.py:217]: This error message is rather specific. We may want to catch and re-raise it at higher levels of the library with a more helpful message.

File: jaxopt/_src/implicit_diff.py [hunk 2]
```diff
@@ -120,29 +120,124 @@ def root_jvp(optimality_fun: Callable,
   return solve(matvec, Jv)
 
 
-def _custom_root(solver_fun, optimality_fun, solve, has_aux):
-  def solver_fun_fwd(init_params, *args):
-    res = solver_fun(init_params, *args)
-    return res, (res, args)
+def _extract_kwargs(kwarg_keys, args):
```
Comment (mblondel) [inline jaxopt/_src/implicit_diff.py:123]: Maybe rename `args` above to `flat_args` (which is consistent with below) or `all_args`. I wasn't sure in the beginning if `args` referred to positional-only arguments or to all arguments.

**Review by mblondel — APPROVED** [review]
Great work, thanks @froystig!

File: jaxopt/_src/implicit_diff.py (hunk 2 above)
Comment (froystig) [reply jaxopt/_src/implicit_diff.py:123]: Good call. Done.

**Pull request merged**